controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,369 @@
+	}
+
+	err = r.ReconcileResource(ctx, &istioclientgoextensionv1alpha1.WasmPlugin{}, desired, rlptools.WASMPluginMutator)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Rate limiting WASMPlugin reconciled successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) desiredRateLimitingWASMPlugin(ctx context.Context, gw *gatewayapiv1.Gateway) (*istioclientgoextensionv1alpha1.WasmPlugin, error) {
+	wasmPlugin := &istioclientgoextensionv1alpha1.WasmPlugin{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "WasmPlugin",
+			APIVersion: "extensions.istio.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rlptools.WASMPluginName(gw),
+			Namespace: gw.Namespace,
+		},
+		Spec: istioextensionsv1alpha1.WasmPlugin{
+			Selector:     common.IstioWorkloadSelectorFromGateway(ctx, r.Client(), gw),
+			Url:          rlptools.WASMFilterImageURL,
+			PluginConfig: nil,
+			// Insert plugin before Istio stats filters and after Istio authorization filters.
+			Phase: istioextensionsv1alpha1.PluginPhase_STATS,
+		},
+	}
+
+	pluginConfig, err := r.wasmPluginConfig(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	if pluginConfig == nil || len(pluginConfig.RateLimitPolicies) == 0 {
+		common.TagObjectToDelete(wasmPlugin)
+		return wasmPlugin, nil
+	}
+
+	pluginConfigStruct, err := pluginConfig.ToStruct()
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin.Spec.PluginConfig = pluginConfigStruct
+
+	// controller reference
+	if err := r.SetOwnerReference(gw, wasmPlugin); err != nil {
+		return nil, err
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) wasmPluginConfig(ctx context.Context, gw *gatewayapiv1.Gateway) (*wasm.Plugin, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &wasm.Plugin{
+		FailureMode:       wasm.FailureModeDeny,
+		RateLimitPolicies: make([]wasm.RateLimitPolicy, 0),
+	}
+
+	gatewayAPITopology, err := r.gatewayAPITopologyFromGateway(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	rateLimitPolicies := gatewayAPITopology.PoliciesFromGateway(gw)
+
+	logger.V(1).Info("wasmPluginConfig", "#RLPS", len(rateLimitPolicies))
+
+	// Shallow copy by assignment
+	rateLimitPoliciesSorted := rateLimitPolicies
+
+	// Sort RLPs for consistent comparison with existing objects
+	sort.Sort(common.PolicyByKey(rateLimitPoliciesSorted))
+
+	for _, policy := range rateLimitPolicies {
+		rlp := policy.(*kuadrantv1beta2.RateLimitPolicy)
+		wasmRLP, err := r.WASMRateLimitPolicy(ctx, gatewayAPITopology, rlp, gw)
+		if err != nil {
+			return nil, err
+		}
+
+		if wasmRLP == nil {
+			// skip this RLP
+			continue
+		}
+
+		wasmPlugin.RateLimitPolicies = append(wasmPlugin.RateLimitPolicies, *wasmRLP)
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) gatewayAPITopologyFromGateway(ctx context.Context, gw *gatewayapiv1.Gateway) (*common.KuadrantTopology, error) {
```

Comment:
Wondering if this func could be enhanced for more kinds of policies, maybe using generics. The fact that it fetches a `RateLimitPolicyList`, which then it iterates through to cast each `RateLimitPolicy` to `KuadrantPolicy`, seems to be a minor detail. WDYT?

Reply:
I agree with Gui's comment, we could benefit from using generics and retrieving a `PolicyList`. I understand we will be requiring it heavily

Reply:
The current implementation of the "Kuadrant Topology" does not use a k8s client to query Gateway API and kuadrant policies. Instead, fetching resources must be done externally and then the kuadrant topology accepts list of resources. It could be done otherwise and have generic function with a k8s client to build topology for a given policy of one kind (for example RLP or KAP). The main reason for the current implementation is to decouple kuadrant topology and the cluster. Meaning that I can build kuadrant topology from a set of resources filtered with some criteria. The usecase used here is that I want to build a kuadrant topology only for one gateway and all policies and routes involved. But only for one gateway. 

Interesting. Worth exploring in future PRs.

Reply:
I guess I'm all right with the `KuadrantTopology` being just a data structure, decoupled from the mechanisms to communicate with the cluster (e.g. API client.)

I still don't see this function as belonging to the WasmPlugin reconciler nevertheless. Despite the RLP kind hard-coded in the core of the function's (current) implementation, it's surrounded by generic code and, in the end, returns something that is not RLP-specific. It feels like the specific policy kind should be a parameter of the function.